sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/FastestServerFetcher.kt

```diff
@@ -0,0 +1,229 @@
+        servers: List<LightWalletEndpoint>,
+    ): Flow<FastestServersResult> {
+        return flow {
+            emit(FastestServersResult(servers = null, isLoading = true))
+
+            val serversByRpcMeanLatency =
+                servers
+                    .parallelMapNotNull {
+                        validateServerEndpointAndMeasure(context, it)
+                    }
+                    .sortedBy {
+                        it.meanDuration
+                    }
+                    .mapIndexedNotNull { index, result ->
+                        if (index <= K - 1 || result.meanDuration <= LATENCY_THRESHOLD) {
+                            Twig.debug { "Fastest Server: '${result.endpoint}' VALIDATED by SORTING by RPC latency" }
+                            result
+                        } else {
+                            Twig.debug { "Fastest Server: '${result.endpoint}' RULED OUT by SORTING by RPC latency" }
+                            null
+                        }
+                    }
+
+            Twig.debug {
+                "Fastest Server: '${serversByRpcMeanLatency.map { it.endpoint }}' VALIDATED by MEASURING RPC latency"
+            }
+
+            emit(FastestServersResult(servers = serversByRpcMeanLatency.map { it.endpoint }.take(K), isLoading = true))
+
+            val serversByGetBlockRangeTimeout =
+                serversByRpcMeanLatency
+                    .asFlow()
+                    .mapNotNull { result ->
+                        val didTimeOut =
+                            withTimeoutOrNull(FETCH_THRESHOLD) {
+                                runCatching {
+                                    val to = result.remoteInfo.blockHeightUnsafe
+                                    val from = BlockHeightUnsafe((to.value - N).coerceAtLeast(0))
+                                    result.lightWalletClient.getBlockRange(from..to)
+                                }.getOrNull()
+                            } == null
+
+                        if (didTimeOut) {
+                            Twig.debug { "Fastest Server: '${result.endpoint}' RULED OUT by getBlockRange timeout" }
+                            null
+                        } else {
+                            Twig.debug { "Fastest Server: '${result.endpoint}' VALIDATED by getBlockRange timeout" }
+                            result
+                        }
+                    }
+                    .map { it.endpoint }
+                    .take(K)
+                    .toList()
+
+            Twig.debug { "Fastest Server: '$serversByGetBlockRangeTimeout' VALIDATED by getBlockRange timeout" }
+
+            emit(FastestServersResult(servers = serversByGetBlockRangeTimeout, isLoading = false))
+        }
+    }
+
+    @Suppress("LongMethod", "ReturnCount")
+    private suspend fun validateServerEndpointAndMeasure(
+        context: Context,
+        endpoint: LightWalletEndpoint
+    ): ValidateServerResult? {
+        fun logRuledOut() {
```

Comment:
Could we add e.g. `reason: String` parameter to this function and provide the reason from its caller? Once the log occurs, we will know which part triggered it.